Support shm sharing

wrvsrx opened the pull request:
Fix #455

wrvsrx commented:
I have tested this PR with [wemeet-wayland-screenshare](https://github.com/xuwd1/wemeet-wayland-screenshare) and [Kooha](https://github.com/SeaDve/Kooha).

It works well with wemeet-wayland-screenshare while kooha's flushing will take a long time, but it's an [issue](https://github.com/SeaDve/Kooha) of kooha

YaLTeR reviewed (changes requested):
Thanks. Please rebase to main.

Haven't looked at the logic in detail yet.

wrvsrx commented:
I‘ll update the PR.

wrvsrx committed:
pw_utils: create a struct called DmaNegotiationResult

wrvsrx committed:
pw_utils: add comments about pw negotiating

wrvsrx committed:
pw_utils: add some comments about `extra_negotiation_result`

wrvsrx committed:
pw_utils: support add and remove shm buffer

wrvsrx committed:
pw_utils: support clean shm buffer

wrvsrx committed:
pw_utils: change paramater of clear_shmbuf to reference

wrvsrx committed:
pw_utils: add `queue_after_sync` to SHM sharing

wrvsrx committed:
pw_utils: rename `mark_buffer_as_good` to `mark_buffer_after_render`

I will change semantic of this function in later commit.

wrvsrx committed:
pw_utils: update chunk info after render to shm buffer

wrvsrx commented on src/pw_utils.rs:
```diff
@@ -31,23 +31,28 @@ use pipewire::stream::{Stream, StreamFlags, StreamListener, StreamState};
 use crate::utils::get_monotonic_time;
 
+const SHM_BLOCKS: usize = 1;
+const SHM_BYTES_PER_PIXEL: usize = 4; // 4 bytes per pixel
```
Done.

wrvsrx committed:
pw_utils: avoid excessive imports

wrvsrx commented on src/pw_utils.rs:
```diff
@@ -963,83 +1163,138 @@ impl CastState {
-    } else {
-        Fourcc::Xrgb8888
-    };
+    pipewire::spa::pod::Object {
```
Done.

wrvsrx commented on src/pw_utils.rs:
```diff
@@ -1105,9 +1360,42 @@ fn allocate_dmabuf(
+
+#[derive(Debug, Clone)]
+pub struct ShmBuffer {
+    fd: Rc<smithay::reexports::rustix::fd::OwnedFd>,
```
Done.

wrvsrx committed:
pw_utils: clean a variable name

wrvsrx commented on src/pw_utils.rs:
```diff
@@ -963,83 +1163,138 @@ impl CastState {
+    refresh: u32,
+    alpha: bool,
+) -> Vec<pod::Object> {
+    let f = |alpha_| {
```
Done.

wrvsrx commented on src/pw_utils.rs:
```diff
@@ -111,24 +115,15 @@ pub enum CastSizeChange {
-            &mut p1[..]
-        };
+        let $params = make_video_params_for_negotiation($formats, $size, $refresh, $alpha);
+        let mut obj_with_buffer: Vec<(&pod::Object, Vec<u8>)> =
+            $params.iter().map(|obj| (obj, Vec::new())).collect();
+        let $params: Vec<_> = obj_with_buffer
+            .iter_mut()
+            .map(|(obj, buf)| make_pod(buf, (*obj).clone()))
+            .collect();
```
I was wondering if you can tell me how to avoid reallocations. I'm not familiar with rust memory model.

wrvsrx commented on src/pw_utils.rs:
```diff
@@ -963,83 +1163,138 @@ impl CastState {
+    size: Size<u32, Physical>,
+    refresh: u32,
+    alpha: bool,
+) -> Vec<pod::Object> {
```
Could you please provide more detailed advice? I'll improve it.

wrvsrx committed:
pw_utils: remove a wrong syncpoint

wrvsrx committed:
pw_utils: avoid some repeated reallocations

wrvsrx commented:
For niri users who use this PR in 25.08, you can use following patch in 25.11.

<https://github.com/wrvsrx/niri/compare/tag_support-shm-sharing_2~19..tag_support-shm-sharing_2.patch>

wrvsrx commented:
> @wrvsrx
> 
> I'm having trouble with many things crashing and going slow on the 25.11 patch. Did you just use the 25.11 tag?

I use 25.11 with my patch, it works well. This patch only changes the behavior when screencasting, can you give an example?

YaLTeR commented:
I merged metadata cursor (changes to pw_utils.rs), then moved pw_utils.rs to src/screencasting/, do you mind rebasing?

wrvsrx commented:
I'll rebase this PR when I get a chance.

superzanti commented:
> > @wrvsrx
> > I'm having trouble with many things crashing and going slow on the 25.11 patch. Did you just use the 25.11 tag?
> 
> I use 25.11 with my patch, it works well. This patch only changes the behavior when screencasting, can you give an example?

So sorry for the confusion. This was just user error.
Seems to work great.

fishman commented:
I just tested the 25.11 patch. It fixes the #2254 screensharing issue in Feishu. It might fix other electron based screensharing issues as well.

wrvsrx committed:
pw_utils: refactor modifier negotiation logic

This commit is preparation for supporting shared memory sharing

wrvsrx committed:
pw_utils: change type of `extra_negotiation_result` to `Option<DmaNegotiationResult>`

This commit is preparation for supporting shared memory sharing

wrvsrx committed:
pw_utils: support sending SHM sharing as fallback when first negotiating

wrvsrx committed:
pw_utils: support negotiating to SHM sharing after connect

wrvsrx commented:
@YaLTeR Rebase has been completed.

wrvsrx committed:
pw_utils: refactor `dequeue_buffer_and_render`

wrvsrx committed:
pw_utils: support render to shm buffer

PurpleCloudX committed:
pw_utils: fixate non-final modifier choices

Initial negotiation can advertise multiple DMA-BUF modifiers. Marking that enum as already fixated makes PipeWire select its default before niri has tested whether GBM can allocate it, which can disconnect otherwise compatible consumers such as OBS.

Keep DONT_FIXATE set while multiple alternatives remain so niri can test and confirm a usable modifier. Preserve the existing implicit-modifier behavior for a lone DRM_FORMAT_MOD_INVALID, and never reopen a choice that niri has already confirmed.

Cover empty, explicit, implicit, multi-modifier, and confirmed choices with unit tests.

PurpleCloudX committed:
pw_utils: validate SHM buffer layout for SPA

SHM allocation previously kept its stride and size as usize values and converted them only when filling SPA metadata. A sufficiently large virtual output could therefore truncate the advertised max size or stride even though the memfd allocation itself succeeded.

Represent the layout using the exact i32 stride and u32 size types required by spa_chunk, and derive memfd allocation, rendering bounds, clearing, and metadata from that single validated value. Reject arithmetic overflow and layouts outside SPA limits before allocating a buffer.

Add coverage for a normal 4K layout, an i32 stride overflow, and a u32 buffer-size overflow.

PurpleCloudX committed:
pw_utils: report complete SHM chunks after rendering

spa_chunk.size describes the valid byte range for MemFd buffers. Reusing the DMA-BUF sentinel value of one therefore exposes only a single byte to consumers that honor SPA metadata, leaving them with empty frames even though niri rendered the full image.

Publish the complete validated SHM layout after rendering, reset the offset and stride, and clear any corruption flag left when the same PipeWire buffer was previously returned unused. Keep the DMA-BUF size sentinel unchanged for compatibility with OBS.

Exercise the update against an actual spa_chunk initialized with stale and corrupted metadata.

wrvsrx committed:
Merge pull request #1 from PurpleCloudX/fix-shm-screencast

pw_utils: fix DMA-BUF negotiation and SHM chunk metadata

bk304 commented:
For some reason beyond comprehension, PipeWire fails to deliver the screen view via DMA-BUF in OpenRGB Effects Plugin's ambient light: "no more input format".

This PR fix this issue. Very nice.

Jivvei commented:
Hi @YaLTeR, are there any plans to merge this PR? Thanks!

Sempyos commented:
YaLTeR is very busy and there is no timeline for features and reviews

YaLTeR commented:
This PR is somewhere at the top of the list of PRs I want to review once I get back to reviewing big PRs.

But I'd also like to remind people that DMA-BUF not working is actually a bug in [PipeWire's GStreamer source](https://gitlab.freedesktop.org/pipewire/pipewire/-/work_items/5353), and this PR makes it work via a slow, inefficient workaround. Which is not to say that shm sharing isn't useful, just that it's not the good final solution for those apps.
